local_settings.sh.in

```diff
@@ -74,40 +127,19 @@ set_port()
 }
 
 # usage: set_file_context a|d (add/delete)
-set_file_contexts()
-{
-	INPUT="fcontext -N -$1 -t named_zone_t \"$SHAREDSTATEDIR/designate/bind9(/.*)?\"
-	fcontext -N -$1 -t virt_cache_t \"$SHAREDSTATEDIR/vhost_sockets(/.*)?\"
-	fcontext -N -$1 -t httpd_var_lib_t $SHAREDSTATEDIR/openstack-dashboard
-	fcontext -N -$1 -t mongod_var_lib_t \"$SHAREDSTATEDIR/mongodb(/.*)?\"
-	fcontext -N -$1 -t httpd_log_t $LOCALSTATEDIR/log/gnocchi/app.log
-	fcontext -N -$1 -t httpd_log_t $LOCALSTATEDIR/log/aodh/app.log
-	fcontext -N -$1 -t httpd_log_t $LOCALSTATEDIR/log/ceilometer/app.log
-	fcontext -N -$1 -t httpd_log_t $LOCALSTATEDIR/log/panko/app.log
-	fcontext -N -$1 -t httpd_log_t $LOCALSTATEDIR/log/zaqar/zaqar.log
-	fcontext -N -$1 -t container_file_t \"$LOCALSTATEDIR/log/containers(/.*)?\"
-	fcontext -N -$1 -t container_file_t \"$LOCALSTATEDIR/lib/config-data(/.*)?\"
-	fcontext -N -$1 -t container_file_t \"$LOCALSTATEDIR/lib/tripleo-config(/.*)?\"
-	fcontext -N -$1 -t neutron_exec_t $BINDIR/neutron-rootwrap-daemon
-	fcontext -N -$1 -t neutron_exec_t $BINDIR/neutron-vpn-agent
-	fcontext -N -$1 -t swift_var_cache_t \"$LOCALSTATEDIR/cache/swift(/.*)\"
-	fcontext -N -$1 -t swift_exec_t $BINDIR/swift-object-reconstructor
-	fcontext -N -$1 -t swift_exec_t $BINDIR/swift-object-relinker
-	fcontext -N -$1 -t httpd_sys_content_t \"${ROOTDIR}httpboot(/.*)?\"
-	fcontext -N -$1 -t ssh_home_t \"$SHAREDSTATEDIR/nova/.ssh(/.*)?\"
-	fcontext -N -$1 -t tftpdir_t \"${ROOTDIR}tftpboot(/.*)?\"
-	fcontext -N -$1 -t cluster_var_log_t \"$LOCALSTATEDIR/log/pacemaker\.log.*\"
-	fcontext -N -$1 -t cluster_var_log_t \"$LOCALSTATEDIR/log/pacemaker(/.*)?\""
-
-	# Load these one by one so upgrades work properly.
-	# TODO (future): Make upgrades (only) do one by one;
-	#                install/remove can do batches to save time.
-	while read; do
-		eval semanage $REPLY &> /dev/null
-	done < <(echo "$INPUT")
-
-	# TODO (future): install/remove can do this to save time
-	# echo "$INPUT" | $SBINDIR/semanage import -N
+set_file_contexts() {
+  case $1 in
+    'a')
+      for f_context in "${!diff_fcontext[@]}"; do
+        eval semanage fcontext -N -${1} -t ${diff_fcontext[$i]} "${f_context}" &> /dev/null
```

Comment:
There are few problems with this line. (I'm still testing on RHEL 8.2)

1. The rules don't seem to apply at all in the current state:
```
./local_settings.sh: eval: line 149: `semanage fcontext -N -a -t tftpdir_t /var/lib/designate/bind9(/.*)?'
./local_settings.sh: eval: line 149: syntax error near unexpected token `('
./local_settings.sh: eval: line 149: `semanage fcontext -N -a -t tftpdir_t /var/log/containers(/.*)?'
ValueError: File context for /usr/bin/neutron-vpn-agent already defined
./local_settings.sh: eval: line 149: syntax error near unexpected token `('
./local_settings.sh: eval: line 149: `semanage fcontext -N -a -t tftpdir_t /var/lib/tripleo-config(/.*)?'
./local_settings.sh: eval: line 149: syntax error near unexpected token `('
./local_settings.sh: eval: line 149: `semanage fcontext -N -a -t tftpdir_t /var/lib/vhost_sockets(/.*)?'
./local_settings.sh: eval: line 149: syntax error near unexpected token `('

```
It looks like the double-quotes need to be escaped. I added that, and the rules applied.

2. Once escaped, the behaviour got weird.

Like, very, very bad at first: every directory got the same one type applied, which is wrong. With an echo added:
```
semanage fcontext -N -a -t tftpdir_t "/var/log/aodh/app.log" &> /dev/null
semanage fcontext -N -a -t tftpdir_t "/var/log/zaqar/zaqar.log" &> /dev/null
semanage fcontext -N -a -t tftpdir_t "/var/log/gnocchi/app.log" &> /dev/null
semanage fcontext -N -a -t tftpdir_t "/httpboot(/.*)?" &> /dev/null
semanage fcontext -N -a -t tftpdir_t "/var/log/pacemaker.log.*" &> /dev/null
```
Everything is getting the tftpdir_t type. Where does $i come from? I tested a few more things and now $i doesn't show anything at all, though at least that means we don't do anything rather than apply the wrong type.

```
semanage fcontext -N -a -t  /var/log/aodh/app.log &> /dev/null
file_spec option is needed for add
```

Not sure if it's something in my environment causing the different behaviour, but there is some kind of issue with this line.

3. The actual output was more like this and I wonder:
```
semanage fcontext -N -a -t tftpdir_t "/var/log/aodh/app.log" &> /dev/null
ValueError: File context for /var/log/aodh/app.log already defined
semanage fcontext -N -a -t tftpdir_t "/var/log/zaqar/zaqar.log" &> /dev/null
ValueError: File context for /var/log/zaqar/zaqar.log already defined
semanage fcontext -N -a -t tftpdir_t "/var/log/gnocchi/app.log" &> /dev/null
ValueError: File context for /var/log/gnocchi/app.log already defined
semanage fcontext -N -a -t tftpdir_t "/httpboot(/.*)?" &> /dev/null
semanage fcontext -N -a -t tftpdir_t "/var/log/pacemaker.log.*" &> /dev/null
ValueError: File context for /var/log/pacemaker.log.* already defined
```

I wonder about the ValueError for already defined fcontexts? They're harmless, but I think we shouldn't be reapplying predefined rules thanks to this patch?
Looking at the "semanage fcontext -Cln" output, pacemaker etc don't show in there either, so that's probably why. Strange.